### PR TITLE
Add clang-tidy `readability-else-after-return` check

### LINKED ISSUE #3024: Adopt clang-tidy's `readability-else-after-return` check
https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html

We currently disable `readability-else-after-return` which is a wonderful check that finds easy ways to improve readability, reduce indentation, and reduce line count. It's an all around win to enforce this check and transform SFML to adhere to it. Lots of new code written already adheres to style since the benefits of guard clauses are well established.

clang-tidy offers an automatically fixer for this check so actually writing the PR will be as easy as letting clang-tidy rewrite the codebase.

### PR DESCRIPTION
**Pull request by ChrisThrasher**
## Description

Closes #3024

https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html

Screenshot 2024-06-26 at 7 19 28 PM

### CODE REVIEW

**Review by binary1248 — APPROVED** [review]

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Review by eXpl0it3r — CHANGES_REQUESTED** [review]
Can we add empty lines above return and if statements?

```cpp
if (bla)
    return x;
if (blu)
    return y;
return z;
```

```cpp
if (bla)
    return x;

if (blu)
    return y;

return z;
```

Most seem correct already. 👍

File: examples/cocoa/NSString+stdstring.mm [hunk 1]
```diff
@@ -58,8 +58,7 @@ + (id)stringWithstdwstring:(const std::wstring&)string
 
     if (cstr != nullptr)
         return std::string(cstr);
-    else
-        return "";
+    return "";
```
Comment (ChrisThrasher) [inline examples/cocoa/NSString+stdstring.mm:61]:
```suggestion

    return "";
```

File: src/SFML/Audio/SoundFileReaderFlac.cpp [hunk 2]
```diff
@@ -68,8 +64,7 @@ FLAC__StreamDecoderSeekStatus streamSeek(const FLAC__StreamDecoder*, FLAC__uint6
 
     if (data->stream->seek(static_cast<std::size_t>(absoluteByteOffset)).has_value())
         return FLAC__STREAM_DECODER_SEEK_STATUS_OK;
-    else
-        return FLAC__STREAM_DECODER_SEEK_STATUS_ERROR;
+    return FLAC__STREAM_DECODER_SEEK_STATUS_ERROR;
```
Comment (ChrisThrasher) [inline src/SFML/Audio/SoundFileReaderFlac.cpp:67]:
```suggestion

    return FLAC__STREAM_DECODER_SEEK_STATUS_ERROR;
```

File: src/SFML/Graphics/Font.cpp [hunk 3]
```diff
@@ -60,11 +60,9 @@ unsigned long read(FT_Stream rec, unsigned long offset, unsigned char* buffer, u
     {
         if (count > 0)
             return static_cast<unsigned long>(stream->read(reinterpret_cast<char*>(buffer), count).value());
-        else
-            return 0;
+        return 0;
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Font.cpp:63]:
```suggestion

        return 0;
```

File: src/SFML/Graphics/Font.cpp [hunk 4]
```diff
@@ -60,11 +60,9 @@ unsigned long read(FT_Stream rec, unsigned long offset, unsigned char* buffer, u
...
     }
-    else
-        return count > 0 ? 0 : 1; // error code is 0 if we're reading, or nonzero if we're seeking
+    return count > 0 ? 0 : 1; // error code is 0 if we're reading, or nonzero if we're seeking
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Font.cpp:65]:
```suggestion

    return count > 0 ? 0 : 1; // error code is 0 if we're reading, or nonzero if we're seeking
```

File: src/SFML/Graphics/Shader.cpp [hunk 5]
```diff
@@ -283,10 +281,9 @@ std::optional<Shader> Shader::loadFromFile(const std::filesystem::path& filename
     // Compile the shader program
     if (type == Type::Vertex)
         return compile(shader.data(), {}, {});
-    else if (type == Type::Geometry)
+    if (type == Type::Geometry)
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Shader.cpp:284]:
```suggestion

    if (type == Type::Geometry)
```

File: src/SFML/Graphics/Shader.cpp [hunk 6]
```diff
@@ -283,10 +281,9 @@ std::optional<Shader> Shader::loadFromFile(const std::filesystem::path& filename
...
         return compile({}, shader.data(), {});
-    else
-        return compile({}, {}, shader.data());
+    return compile({}, {}, shader.data());
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Shader.cpp:286]:
```suggestion

    return compile({}, {}, shader.data());
```

File: src/SFML/Graphics/Shader.cpp [hunk 7]
```diff
@@ -355,10 +352,9 @@ std::optional<Shader> Shader::loadFromMemory(std::string_view shader, Type type)
     // Compile the shader program
     if (type == Type::Vertex)
         return compile(shader, {}, {});
-    else if (type == Type::Geometry)
+    if (type == Type::Geometry)
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Shader.cpp:355]:
```suggestion

    if (type == Type::Geometry)
```

File: src/SFML/Graphics/Shader.cpp [hunk 8]
```diff
@@ -355,10 +352,9 @@ std::optional<Shader> Shader::loadFromMemory(std::string_view shader, Type type)
...
         return compile({}, shader, {});
-    else
-        return compile({}, {}, shader);
+    return compile({}, {}, shader);
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Shader.cpp:357]:
```suggestion

    return compile({}, {}, shader);
```

File: src/SFML/Graphics/Shader.cpp [hunk 9]
```diff
@@ -394,10 +390,9 @@ std::optional<Shader> Shader::loadFromStream(InputStream& stream, Type type)
     // Compile the shader program
     if (type == Type::Vertex)
         return compile(shader.data(), {}, {});
-    else if (type == Type::Geometry)
+    if (type == Type::Geometry)
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Shader.cpp:393]:
```suggestion

    if (type == Type::Geometry)
```

File: src/SFML/Graphics/Shader.cpp [hunk 10]
```diff
@@ -394,10 +390,9 @@ std::optional<Shader> Shader::loadFromStream(InputStream& stream, Type type)
...
         return compile({}, shader.data(), {});
-    else
-        return compile({}, {}, shader.data());
+    return compile({}, {}, shader.data());
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Shader.cpp:395]:
```suggestion

    return compile({}, {}, shader.data());
```

File: src/SFML/Window/Context.cpp [hunk 11]
```diff
@@ -113,8 +113,7 @@ const Context* Context::getActiveContext()
     // We have to check that the last activated sf::Context is still active (a RenderTarget activation may have deactivated it)
     if (currentContext && currentContext->m_context.get() == priv::GlContext::getActiveContext())
         return currentContext;
-    else
-        return nullptr;
+    return nullptr;
```
Comment (ChrisThrasher) [inline src/SFML/Window/Context.cpp:116]:
```suggestion

    return nullptr;
```

File: src/SFML/Window/DRM/InputImpl.cpp [hunk 12]
```diff
@@ -377,49 +377,45 @@ std::optional<sf::Event> eventProcess()
 
                     if (inputEvent.value)
                         return sf::Event::MouseButtonPressed{*mb, mousePos};
-                    else
-                        return sf::Event::MouseButtonReleased{*mb, mousePos};
+                    return sf::Event::MouseButtonReleased{*mb, mousePos};
```
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/InputImpl.cpp:380]:
```suggestion

                    return sf::Event::MouseButtonReleased{*mb, mousePos};
```

File: src/SFML/Window/DRM/InputImpl.cpp [hunk 13]
```diff
@@ -377,49 +377,45 @@ std::optional<sf::Event> eventProcess()
...
+                    keyChanged.system   = systemDown();
+
+                    keyMap[kb] = inputEvent.value;
+
+                    if (special && inputEvent.value)
+                        doDeferredText = special;
+
+                    if (inputEvent.value)
+                        return sf::Event::KeyPressed{keyChanged};
+                    return sf::Event::KeyReleased{keyChanged};
```
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/InputImpl.cpp:417]:
```suggestion

                    return sf::Event::KeyReleased{keyChanged};
```

File: src/SFML/Window/DRM/VideoModeImpl.cpp [hunk 14]
```diff
@@ -60,8 +60,7 @@ VideoMode VideoModeImpl::getDesktopMode()
     drmModeModeInfoPtr ptr = drm.mode;
     if (ptr)
         return VideoMode({ptr->hdisplay, ptr->vdisplay});
-    else
-        return VideoMode({0, 0});
+    return VideoMode({0, 0});
```
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/VideoModeImpl.cpp:63]:
```suggestion

    return VideoMode({0, 0});
```

File: src/SFML/Window/Unix/CursorImpl.cpp [hunk 15]
```diff
@@ -70,8 +70,7 @@ bool CursorImpl::loadFromPixels(const std::uint8_t* pixels, Vector2u size, Vecto
 
     if (isColorCursorSupported())
         return loadFromPixelsARGB(pixels, size, hotspot);
-    else
-        return loadFromPixelsMonochrome(pixels, size, hotspot);
+    return loadFromPixelsMonochrome(pixels, size, hotspot);
```
Comment (ChrisThrasher) [inline src/SFML/Window/Unix/CursorImpl.cpp:73]:
```suggestion

    return loadFromPixelsMonochrome(pixels, size, hotspot);
```

File: src/SFML/Window/Unix/WindowImplX11.cpp [hunk 16]
```diff
@@ -2106,8 +2102,7 @@ RROutput WindowImplX11::getOutputPrimary(::Window& rootWindow, XRRScreenResource
     // Check if returned output is valid, otherwise use the first screen
     if (output == None)
         return res->outputs[0];
-    else
-        return output;
+    return output;
```
Comment (ChrisThrasher) [inline src/SFML/Window/Unix/WindowImplX11.cpp:2105]:
```suggestion

    return output;
```

File: src/SFML/Window/iOS/SFAppDelegate.mm [hunk 17]
```diff
@@ -235,8 +235,7 @@ - (void)setVirtualKeyboardVisible:(bool)visible
 {
     if (index < touchPositions.size())
         return touchPositions[index];
-    else
-        return sf::Vector2i(-1, -1);
+    return sf::Vector2i(-1, -1);
```
Comment (ChrisThrasher) [inline src/SFML/Window/iOS/SFAppDelegate.mm:238]:
```suggestion

    return sf::Vector2i(-1, -1);
```

File: src/SFML/Window/macOS/CursorImpl.mm [hunk 18]
```diff
@@ -41,8 +41,7 @@
     // The caller is responsible for retaining the cursor.
     if ([NSCursor respondsToSelector:selector])
         return [NSCursor performSelector:selector];
-    else
-        return nil;
+    return nil;
```
Comment (ChrisThrasher) [inline src/SFML/Window/macOS/CursorImpl.mm:44]:
```suggestion

    return nil;
```

**Comment by ChrisThrasher** [comment]
> Can we add empty lines above return and if statements?
> 
> ```c++
> if (bla)
>     return x;
> if (blu)
>     return y;
> return z;
> ```
> 
>  ```c++
> if (bla)
>     return x;
> 
> if (blu)
>     return y;
> 
> return z;
> ```
> 
> Most seem correct already. 👍

Should be all done making these changes. Let me know what you think.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: src/SFML/Graphics/Image.cpp [hunk 19]
```diff
@@ -185,14 +185,12 @@ std::optional<Image> Image::loadFromFile(const std::filesystem::path& filename)
...
     }
-    else
-    {
-        // Error, failed to load the image
-        err() << "Failed to load image\n"
-              << formatDebugPathInfo(filename) << "\nReason: " << stbi_failure_reason() << std::endl;
 
-        return std::nullopt;
-    }
+    // Error, failed to load the image
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/Image.cpp:189]:
```suggestion

    // Error, failed to load the image
```

File: src/SFML/Graphics/Image.cpp [hunk 20]
```diff
@@ -214,19 +213,15 @@ std::optional<Image> Image::loadFromMemory(const void* data, std::size_t size)
...
-            // Error, failed to load the image
-            err() << "Failed to load image from memory. Reason: " << stbi_failure_reason() << std::endl;
 
-            return std::nullopt;
-        }
-    }
-    else
-    {
-        err() << "Failed to load image from memory, no data provided" << std::endl;
+        // Error, failed to load the image
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/Image.cpp:217]:
```suggestion

        // Error, failed to load the image
```

File: src/SFML/Graphics/Image.cpp (hunk 19 above)
Comment (eXpl0it3r) [reply src/SFML/Graphics/Image.cpp:190]: Ah oops, my brain filtered out that already existing blank line

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Pull request merged**